### PR TITLE
Tabs revisited

### PR DESCRIPTION
**Pull request by jonathanadler**
new architecture.
Need to add tests and updated documentation

### CODE REVIEW

File: src/components/Tabs/Tabs.tsx [hunk 1]
```diff
@@ -33,169 +36,157 @@ interface TabsState {
...
-              onRightNavClick={this._onNavClickRight}
+              activeTabIndex={activeTabIndex}
+              ref={this._tabsRefCallback}
+              key={tabsKey}
             />
+            <TabsNavButton
+              onClick={this._onClickLeft}
+              className={classnames(style.navBtn, style.navBtnLeft)}
+            >
+              <ChevronLeft size="35" />
```
Comment (shlomitc) [inline src/components/Tabs/Tabs.tsx]: I'd use `className` setting the width and height

File: src/components/Tabs/ScrollableTabs/ScrollableTabs.tsx [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+  }
+
+  render() {
+    const { selectedIndicatorRect } = this.state;
+    const { onScroll, items, alignment, variant, ...rest } = this.props;
+
+    return (
+      <div {...style('root', { alignment, variant }, rest)}>
+        <nav className={style.nav} ref={this._navRef} onScroll={onScroll}>
+          {items.map(this._getTabItem)}
```
Comment (shlomitc) [inline src/components/Tabs/ScrollableTabs/ScrollableTabs.tsx]: prefer setting the tab item inline

File: src/components/Tabs/ScrollableTabs/ScrollableTabs.tsx [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        rightLimit;
+
+      if (leftDelta < 0 || rightDelta > 0) {
+        this._animateScroll(currentTabElement.offsetLeft);
+      }
+    }
+  };
+
+  _animateScroll(scrollLeft: number) {
+    animateProp('scrollLeft', this._navRef.current, scrollLeft);
```
Comment (shlomitc) [inline src/components/Tabs/ScrollableTabs/ScrollableTabs.tsx]: what does `animateProp` mean?

File: src/components/Tabs/ScrollableTabs/ScrollableTabs.tsx [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+  scrollLeft() {
+    this._animateScroll(-1 * this._navRef.current.clientWidth);
+  }
+
+  scrollRight() {
+    this._animateScroll(this._navRef.current.clientWidth);
+  }
+
+  getNavElement() {
```
Comment (shlomitc) [inline src/components/Tabs/ScrollableTabs/ScrollableTabs.tsx]: consider abstracting to the return only the dimensions

File: src/components/Tabs/ScrollableTabs/ScrollableTabs.tsx [hunk 5]
```diff
@@ -0,0 +1,151 @@
+import * as React from 'react';
```
Comment (shlomitc) [inline src/components/Tabs/ScrollableTabs/ScrollableTabs.tsx]: expose an index file

File: src/components/Tabs/Tabs.st.css [hunk 6]
```diff
@@ -65,94 +70,80 @@
...
+.root:navButtons(both) .navigation::nav {
+    left: calc(-1 * value(navBtnWidth));
 }
 
-.root:variant(fullWidth) nav,
-.root:mobile nav {
-    display: flex;
+.root:navButtons(right) .navigation::nav,
+.root:navButtons(both) .navigation::nav {
+    right: calc(-1 * value(navBtnWidth));
```
Comment (shlomitc) [inline src/components/Tabs/Tabs.st.css]: change to static calculate function

File: src/components/Tabs/ScrollableTabs/ScrollableTabs.tsx [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+  };
+
+  constructor(props) {
+    super(props);
+
+    this._navRef = React.createRef();
+    this._selectedTabRef = React.createRef();
+  }
+
+  componentDidMount = this._updateComponent;
```
Comment (shlomitc) [inline src/components/Tabs/ScrollableTabs/ScrollableTabs.tsx]: change to function that calls _updateComponent

File: src/components/Tabs/ScrollableTabs/ScrollableTabs.tsx [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+
+  componentDidUpdate(prevProps: ScrollableTabsProps) {
```
Comment (shlomitc) [inline src/components/Tabs/ScrollableTabs/ScrollableTabs.tsx]: should be tested with `mount`

File: src/components/Tabs/Tab/Tab.tsx [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    }
+  };
+
+  return (
+    <div
+      {...style('root', {}, props)}
+      ref={ref}
+      data-index={index}
+      data-active={isActive}
+      data-hook={`tab-item-${index}`}
```
Comment (shlomitc) [inline src/components/Tabs/Tab/Tab.tsx]: should come from parent

File: src/components/Tabs/Tabs.driver.ts [hunk 10]
```diff
@@ -1,22 +1,27 @@
 import { StylableDOMUtil } from '@stylable/dom-test-kit';
 import style from './Tabs.st.css';
+import innerStyle from './ScrollableTabs/ScrollableTabs.st.css';
```
Comment (shlomitc) [inline src/components/Tabs/Tabs.driver.ts]: prefer not importing any stylesheets in our drivers. use `data-attr`

File: src/components/Tabs/ScrollableTabs/ScrollableTabs.st.css [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    -st-default: Tab;
+}
+
+:vars {
+    defaultSelectedTabIndicatorColor: color-8;
+}
+
+:vars {
+    selectedTabIndicatorColor: --overridable;
+}
```
Comment (NitayRabi) [inline src/components/Tabs/ScrollableTabs/ScrollableTabs.st.css]: Why separate `:vars` ?

File: src/components/Tabs/ScrollableTabs/ScrollableTabs.tsx [hunk 12]
```diff
@@ -0,0 +1,175 @@
+import * as React from 'react';
+import { Tab, TabItem } from '../Tab/Tab';
+import { ALIGNMENT as Alignment, VARIANT as Variant } from '../../Tabs';
+import { animateElementByProp } from '../../../common/animations';
+import style from './ScrollableTabs.st.css';
+
+const isEqual = require('lodash/isEqual');
```
Comment (NitayRabi) [inline src/components/Tabs/ScrollableTabs/ScrollableTabs.tsx]: any special reason for `require` over `import` syntax?

File: src/components/Tabs/Tab/Tab.tsx [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+    <div
+      {...style('root', {}, props)}
+      ref={ref}
+      data-index={index}
+      data-active={isActive}
+      data-hook={dataHook}
+      key={`${title}-${index}`}
+      onClick={onSelectTab}
+      onKeyDown={onKeyDown}
+      tabIndex={0}
```
Comment (NitayRabi) [inline src/components/Tabs/Tab/Tab.tsx]: It make sense to put same tabIndex for the inner tabs (probably the same as the parent or parent + 1) but in storyBook tabs doesn't seem to work - needs to be tested and implemented.

File: src/components/Tabs/Tabs.driver.ts [hunk 14]
```diff
@@ -1,22 +1,22 @@
...
 import { Simulate } from 'react-dom/test-utils';
 
 export const tabsDriverFactory = ({ element }) => {
-  const stylableDOMUtil = new StylableDOMUtil(style);
-
   const getItems: any = () =>
     Array.from(element.querySelector('nav').childNodes);
 
+  const getTabs: any = () =>
+    element.querySelector('[data-hook="scrollable-tabs"]');
```
Comment (NitayRabi) [inline src/components/Tabs/Tabs.driver.ts]: Use constant for dataHooks/data-X names (to be reused in production and testing code)

File: src/components/Tabs/Tabs.st.css [hunk 15]
```diff
@@ -1,10 +1,16 @@
 :import {
-    -st-from: "../../tpaStyles/styleableExt";
+    -st-from: "../../tpaStyles/styleableExt.js";
```
Comment (NitayRabi) [inline src/components/Tabs/Tabs.st.css]: is the `.js` needed (not according to the Stylable docs) https://stylable.io/docs/references/formatters

File: src/components/Tabs/ScrollableTabs/ScrollableTabs.st.css [hunk 16]
```diff
@@ -0,0 +1,65 @@
+:import {
+    -st-from: "../../../tpaStyles/styleableExt.js";
```
Comment (NitayRabi) [inline src/components/Tabs/ScrollableTabs/ScrollableTabs.st.css]: is the .js needed (not according to the Stylable docs)
https://stylable.io/docs/references/formatters

File: src/components/Tabs/Tabs.st.css [hunk 17]
```diff
@@ -24,37 +30,37 @@
     IndicatorColor: --overridable;
 }
 
+:vars {
+    calculatedBorderColor: wixApply(color, wixApply(fallback, value(IndicatorColor), wixApply(opacity, wixApply(color, value(defaultIndicatorColor)), 0.2)));
+}
```
Comment (NitayRabi) [inline src/components/Tabs/Tabs.st.css]: Why separate `:var` declarations?

File: src/components/Tabs/Tabs.st.css [hunk 18]
```diff
@@ -1,10 +1,16 @@
...
 }
 
+:import {
+    -st-from: "./ScrollableTabs/ScrollableTabs.st.css";
+    -st-default: Navigation;
+}
+
 :vars {
     tabsHeight: 52px;
+    tabsMobileHeight: 48px;
```
Comment (NitayRabi) [inline src/components/Tabs/Tabs.st.css]: Theses are hardcoded default values? shouldn't be provided through santa?

File: src/components/Tabs/Tabs.tsx [hunk 19]
```diff
@@ -33,169 +36,141 @@ interface TabsState {
...
               variant={variant}
+              items={items}
+              className={style.navigation}
+              onClickItem={this._onClickItem}
               onScroll={this._onScroll}
-              onLeftNavClick={this._onNavClickLeft}
-              onRightNavClick={this._onNavClickRight}
+              activeTabIndex={activeTabIndex}
+              ref={this._tabsRefCallback}
+              key={tabsKey}
```
Comment (NitayRabi) [inline src/components/Tabs/Tabs.tsx]: If this key is only used for re-rendring `ScrollableTabs`, we might consider a different solution.

File: src/components/Tabs/ScrollableTabs/ScrollableTabs.tsx (hunk 12 above)
Comment (jonathanadler) [reply src/components/Tabs/ScrollableTabs/ScrollableTabs.tsx]: typescript issues. I'll add a @ts-ignore comment

File: src/components/Tabs/Tabs.driver.ts (hunk 14 above)
Comment (jonathanadler) [reply src/components/Tabs/Tabs.driver.ts]: I added a `DATA_HOOKS` constant, but i'm not sure adding the `data-X` names will result in readable code. Can you show me an example maybe?

File: src/components/Tabs/Tabs.st.css (hunk 17 above)
Comment (jonathanadler) [reply src/components/Tabs/Tabs.st.css]: Theres a convention in this project of distinguishing between variables that are overridable and those that aren't.
I've changed it a bit. Soon to be pushed.

File: src/components/Tabs/Tabs.st.css (hunk 18 above)
Comment (jonathanadler) [reply src/components/Tabs/Tabs.st.css]: No, these aren't overridable

File: src/components/Tabs/Tab/Tab.tsx (hunk 13 above)
Comment (jonathanadler) [reply src/components/Tabs/Tab/Tab.tsx]: I'm not sure the parent should be focusable. 
Nevertheless, `tabindex={0}` will work with the natural structure of the component

**Comment by NitayRabi** [comment]
I would break this down separate PRs for storybook-test and tabs

**Pull request merged**